`code_warntype` now shows unstable SSA useref values in red

arhik opened the pull request:
Addresses https://github.com/JuliaLang/julia/issues/54028

reraising previous PR after suggested code changes https://github.com/JuliaLang/julia/pull/54441
![Screenshot 2024-05-17 at 8 16 09 AM](https://github.com/JuliaLang/julia/assets/5755530/5c1ff893-f345-4c6e-ae38-576ceb69f432)

arhik committed:
method to check if a statement of CodeInfo is unstable.

arhik committed:
pass determined unstable_ssa statements

This is partial commit which passes unstable_ssa statements to `show_ir_stmts`.
Central Idea is to show/print type unstable ssa userefs in IR statements in color. To achieve that we need to modify following code segments.
[*] determine unstable_ssa
[*] pass unstable_ssa to `show_ir_stmts`.
[ ] pass unstable_ssa to `show_ir_stmt`.
[ ] pass unstable_ssa to `print_stmt`
[ ] pass unstable_ssa to `show_unquoted`

This PR covers second task alone

arhik committed:
pass each statement unstable_ssa information

This is partial commit which passes unstable_ssa statements to `show_ir_stmts`.
Central Idea is to show/print type unstable ssa userefs in IR statements in color. To achieve that we need to modify following code segments.
[*] determine unstable_ssa
[*] pass unstable_ssa to `show_ir_stmts`.
[*] pass unstable_ssa to `show_ir_stmt`.
[ ] pass unstable_ssa to `print_stmt`
[ ] pass unstable_ssa to `show_unquoted`

This PR covers third task alone

arhik committed:
pass unstable_ssa to print statement

This is partial commit which passes unstable_ssa statements to `show_ir_stmts`.
Central Idea is to show/print type unstable ssa userefs in IR statements in color. To achieve that we need to modify following code segments.
[*] determine unstable_ssa
[*] pass unstable_ssa to `show_ir_stmts`.
[*] pass unstable_ssa to `show_ir_stmt`.
[*] pass unstable_ssa to `print_stmt`
[ ] pass unstable_ssa to `show_unquoted`

This PR covers fourth task alone

arhik committed:
pass unstable_ssa to show_unquoted

This is partial commit which passes unstable_ssa statements to `show_ir_stmts`.
Central Idea is to show/print type unstable ssa userefs in IR statements in color. To achieve that we need to modify following code segments.
[*] determine unstable_ssa
[*] pass unstable_ssa to `show_ir_stmts`.
[*] pass unstable_ssa to `show_ir_stmt`.
[*] pass unstable_ssa to `print_stmt`
[*] pass unstable_ssa to `show_unquoted`

This PR covers finished basic pipeline. Other ssa values appearing in nodes like `return` etc can be easily provided now.

arhik committed:
32bit version compatibility

arhik committed:
unstable_ssa as optional in `show_unquoted`

arhik committed:
code_warntype colored ssa for binary ops

ops like `x + y` are covered.

arhik commented:
Highlighting ssa values on the left is also useful. Not part of this PR but can be done after feedback.
![Screenshot 2024-05-17 at 9 06 56 AM](https://github.com/JuliaLang/julia/assets/5755530/a7f49976-43da-4bb6-bf6d-e9da5b580dcd)

arhik commented:
This is without left highlighting for comparison.

![Screenshot 2024-05-17 at 9 19 52 AM](https://github.com/JuliaLang/julia/assets/5755530/e8c52837-523e-4d93-9ec4-bc8346712877)

arhik committed:
highlighting unstable SSA on the left too

arhik commented:
Does this need tests ?

arhik commented:
@gbaraldi If this needs more discussion let me know.

gbaraldi commented:
It's always good to have some tests yeah

fingolfin commented on base/compiler/ssair/show.jl:
```diff
@@ -104,45 +116,63 @@ function show_unquoted_phinode(io::IO, stmt::PhiNode, indent::Int, prefix::Strin
         first ? (first = false) : print(io, ", ")
-        show_unquoted(io, v, indent)
+        if v isa SSAValue
+            show_unquoted(io, v, indent, prec, (v in unstable_ssa) ? true : false)
```
Option suggestion, but: how about changing this and its many variants in this PR to
```suggestion
            show_unquoted(io, v, indent, prec, unstable_ssa)
```
and then deal with it in the `show_unquoted(::IO, ::SSAValue, ...)` method. 

That would then also make it trivial to use `unstable_ssa::Union{BitSet,Nothing} = nothing` everywhere to avoid creating a `BitSet()` if it is not needed. Dealing with the `nothing` case would only have to be done in one place, namely the `show_unquoted` method for `SSAValue`

arhik replied:
I remember I had to do that to avoid stackoverflow. But I will check again. Reasonable suggestion to invest time. Thanks.

fingolfin commented on base/compiler/ssair/show.jl:
```diff
@@ -89,12 +96,17 @@ end
-                show_unquoted(io′, stmt.values[i], indent)
+                val = stmt.values[i]
+                if val isa SSAValue
+                    show_unquoted(io′, stmt.values[i], indent, 0, (stmt.values[i].id in unstable_ssa) ? true : false)
+                else
+                    show_unquoted(io′, stmt.values[i], indent)
```
```suggestion
                    show_unquoted(io′, val, indent, 0, (stmt.values[i].id in unstable_ssa) ? true : false)
                else
                    show_unquoted(io′, val, indent)
```

or even
```suggestion
                    show_unquoted(io′, val, indent, 0, stmt.values[i].id in unstable_ssa)
                else
                    show_unquoted(io′, val, indent)
```

fingolfin commented on base/compiler/ssair/show.jl:
```diff
@@ -104,45 +116,63 @@ function show_unquoted_phinode(io::IO, stmt::PhiNode, indent::Int, prefix::Strin
+function show_unquoted(io::IO, stmt::PiNode, indent::Int, prec::Int, unstable_ssa::BitSet=BitSet())
     print(io, "π (")
-    show_unquoted(io, stmt.val, indent)
+    show_unquoted(io, stmt.val, indent, prec, (stmt.val.id in unstable_ssa) ? true : false)
```
So previously `prec` was ignored, now it is passed on. Is this a necessary for this patch? (I am not objecting, just wondering whether this fixes something, resp. where it might break something else)

fingolfin reviewed (commented):
Thank you that looks like a quite helpful improvement!

Left some minor and optional suggestions, feel free to ignore/disagree.

arhik committed:
colored ssa fixes in Phi and GoTo nodes

arhik committed:
embracing the fact that `printstyled` with `:normal` color and `print` are different

arhik committed:
Update show.jl

arhik committed:
method to check if a statement of CodeInfo is unstable.

pass unstable_ssa to show_unquoted

This is partial commit which passes unstable_ssa statements to `show_ir_stmts`.
Central Idea is to show/print type unstable ssa userefs in IR statements in color. To achieve that we need to modify following code segments.
[*] determine unstable_ssa
[*] pass unstable_ssa to `show_ir_stmts`.
[*] pass unstable_ssa to `show_ir_stmt`.
[*] pass unstable_ssa to `print_stmt`
[*] pass unstable_ssa to `show_unquoted`

This PR covers finished basic pipeline. Other ssa values appearing in nodes like `return` etc can be easily provided now.

unstable_ssa as optional in `show_unquoted`

unstable ssa should be highlighted only when appropriate
